Add cuda device synchronization support in JIT

nikithamalgifb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55469 Add cuda device synchronization support in JIT**

Differential Revision: [D27749077](https://our.internmc.facebook.com/intern/diff/D27749077)

nikithamalgifb committed:
Add cuda device synchronization support in JIT

[ghstack-poisoned]

nikithamalgifb committed:
Update on "Add cuda device synchronization support in JIT"




[ghstack-poisoned]

SplitInfinity commented on test/jit/test_cuda.py:
```diff
@@ -49,6 +49,35 @@ def tearDown(self):
+            torch.cuda.synchronize(torch.device('cuda:1'))
+            after_current_device_index = torch.cuda.current_device()
+
+            return prev_current_device_index == after_current_device_index
```
Consider adding a comment explaining why this check is needed.

SplitInfinity commented on test/jit/test_cuda.py:
```diff
@@ -49,6 +49,35 @@ def tearDown(self):
+            torch.cuda.synchronize(1)
+            after_current_device_index = torch.cuda.current_device()
+
+            return prev_current_device_index == after_current_device_index
```
Consider adding a comment explaining why this check is needed.

nikithamalgifb replied:
Done

SplitInfinity commented on torch/csrc/jit/runtime/register_cuda_ops.cpp:
```diff
@@ -119,6 +119,54 @@ RegisterOperators const reg({
+    Operator(
+        "cuda::synchronize.device(Device? device) -> ()",
+        [](Stack* stack) {
+          auto device = pop(stack).toOptional<c10::Device>();
+          auto current_device_index = c10::cuda::current_device();
+          c10::DeviceIndex device_index =
+              device.has_value() ? device->index() : current_device_index;
+
+          // If the current_device and the device to synchronize are not
+          // the same, set the device to the device_index of the device
+          // to synchronize.
+          if (current_device_index != device_index) {
+            c10::cuda::set_device(device_index);
+          }
+          c10::cuda::device_synchronize();
+
+          // Reset the device to current_device before synchronizing.
+          if (current_device_index != device_index) {
+            c10::cuda::set_device(current_device_index);
+          }
```
This code looks extremely similar to the implementation for the `int64_t` variant below, consider refactoring.

nikithamalgifb replied:
Done

SplitInfinity reviewed (commented):
Do we need to mark these operators as having side effects so that they don't get DCE'd? 

Consider using FileCheck in the tests to make assertions about what's in the graph. I wouldn't be surprised if the tests in their current form aren't actually calling the ops...

nikithamalgifb committed:
Update on "Add cuda device synchronization support in JIT"




[ghstack-poisoned]

nikithamalgifb commented:
> Do we need to mark these operators as having side effects so that they don't get DCE'd?
> 
> Consider using FileCheck in the tests to make assertions about what's in the graph. I wouldn't be surprised if the tests in their current form aren't actually calling the ops...

Yeah, they were getting DCE'd. I have fixed this now.

nikithamalgifb committed:
Update on "Add cuda device synchronization support in JIT"




[ghstack-poisoned]

SplitInfinity reviewed (approved):
Thanks for addressing the comments I made. I would still highly recommend adding some FileCheck checks to the unit tests to make sure the operator isn't getting DCE'd.

nikithamalgifb commented:
> Thanks for addressing the comments I made. I would still highly recommend adding some FileCheck checks to the unit tests to make sure the operator isn't getting DCE'd.

Thanks Meghan. I have added the checks to synchronize test cases. I'll create a new PR to handle FileChecks for the rest of the tests

SplitInfinity commented:
> > Thanks for addressing the comments I made. I would still highly recommend adding some FileCheck checks to the unit tests to make sure the operator isn't getting DCE'd.
> 
> Thanks. I have added the checks to synchronize test cases. I'll create a new PR to handle FileChecks for the rest of the tests

Oops, I did not see that. Sorry!
